glean/lang/lsif/Data/LSIF/Angle.hs

```diff
@@ -0,0 +1,255 @@
+we might want this typed for maintenance reasons.
+
+-}
+
+{-# LANGUAGE OverloadedStrings #-}
+
+module Data.LSIF.Angle (toAngle) where
+
+-- from
+import Data.LSIF.Types
+import Data.LSIF.JSON ({-instances -})
+import qualified Data.LSIF.Types as LSIF
+-- to
+-- import qualified Glean.Schema.Lsif.Types as Glean
+-- import qualified Glean.Schema.Src.Types as Src
+-- via
+import Data.Aeson ( object, Value(String, Array), KeyValue(..) )
+import Data.Aeson.Types ( Pair )
+import Data.Maybe ( fromMaybe )
+import Data.Text ( Text )
+import Data.Int ( Int64 )
+import qualified Data.Text as Text
+import Data.IntSet ( IntSet )
+import Data.IntMap ( IntMap )
+import qualified Data.IntSet as Set
+import qualified Data.IntMap.Strict as IMap
+
+import qualified Data.Vector as V
+import Control.Monad.State.Strict
+import Control.Monad ( forM )
+
+-- Converts from AESON-parsed standard LSIF into Glean-compatible types
+
+-- | Parser env, to track command statements in LSIF that change scope
+data Env =
+  Env {
+    root :: !Text, -- ^ used to clean up file paths to be repo-relative
+    openProjects :: IntSet, -- open project identifiers
+    openDocuments :: IntSet, -- stack of open document identifiers
+
+    -- "type" environment, track vertex/$typeResult ids for defs/refs/decls
+    typemap :: IntMap SymbolType -- keep track of type of result ids
+  }
+
+data SymbolType
+  = DefinitionType
+  | DeclarationType
+  | ReferenceType
+  deriving (Eq, Ord, Bounded, Enum)
+
+empty :: Env
+empty = Env "" Set.empty Set.empty IMap.empty
+
+-- Drop projectRoot prefix from URI to yield repo-relative paths for Glean
+filterRoot :: Text -> State Env Text
+filterRoot path = do
+  prefix <- root <$> get
+  pure $ fromMaybe path (Text.stripPrefix (prefix <> "/") path)
+
+setRoot :: Text -> State Env ()
+setRoot path = modify' (\e -> e { root = path })
+
+insertProject, insertDocument :: Id -> State Env ()
+insertProject (Id n) =
+  modify' (\e -> e { openProjects = Set.insert (fromIntegral n) (openProjects e) })
+insertDocument (Id n) =
+  modify' (\e -> e { openDocuments = Set.insert (fromIntegral n) (openDocuments e) })
+
+insertType :: Id -> SymbolType -> State Env ()
+insertType (Id n) ty =
+  modify' (\e -> e { typemap = IMap.insert (fromIntegral n) ty (typemap e) })
+
+getTypeOf :: Id -> State Env (Maybe SymbolType)
+getTypeOf (Id n) = do
+  imap <- typemap <$> get
+  pure (IMap.lookup (fromIntegral n) imap)
+
+
+--
+-- | Convert AESON.LSIF to Glean.JSON
+--
+-- We map over the statements, in order, generating 0 or more facts for Glean
+-- LSIF guarantees things are defined before references to them are used, and
+-- uses unique ids for facts. We can use that to simplify processing
+--
+-- vertexes correspond to facts holding new data
+-- edges relate existnig vertex facts to each other (with new facts)
+--
+-- Currently we emit one predicate per fact in LSIF. Which is verbose. We should
```

Comment:
'one Angle fact per LSIF fact' might be clearer. Otherwise it seems like we are creating schema predicates.